serde/de.py

```diff
@@ -569,6 +594,44 @@ def {{func}}(data, reuse_instances = {{reuse_instances_default}}):
     return env.get_template('dict').render(func=FROM_DICT, cls=cls, reuse_instances_default=reuse_instances_default)
 
 
+def render_union_func(cls: Type, union_args: List[Type]) -> str:
+    template = """
+def {{func}}(serde_scope, obj, reuse_instances):
+  {% for name in cls.__serde_scope__ %}
+  {{name}} = serde_scope['{{name}}']
+  {% endfor %}
+
+  # create fake dict so we can reuse the normal render function
+  fake_dict = {"fake_key":obj}
+  
+  error = "Exhausted all types"
+  {% for t in union_args %}
+  {% if t | is_primitive %}
+  if isinstance(obj, {{t.__name__}}):
+    return {{t|arg|rvalue}}
+  else:
+    error = "input is not of type {{t.__name__}}"
+  {% else %}
+  try:
+    return {{t|arg|rvalue}}
+  except Exception as e:
+    error = str(e)
+  {% endif %}
+  {% endfor %}
+  raise SerdeError("Can not deserialize " + repr(obj) + " of type " + typename(type(obj)) + " into {{union_name}}. Reason: " + error)
+    """
```

Comment:
General idea of the implementation looks good to me! 😄  

I had tried to implement Union support back in [pyserde 0.0.7](https://github.com/yukinarit/pyserde/commit/8c4503c31d37bf074381db723ff463ba13a0b88f), but my implementation wasn't generation based. It resolved a matching type in Union at runtime, thus it wasn't efficient enough.